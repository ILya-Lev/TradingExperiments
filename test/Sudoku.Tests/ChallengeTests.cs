using ApprovalTests;
using ApprovalTests.Reporters;
using ApprovalTests.Reporters.TestFrameworks;
using FluentAssertions;
using Xunit.Abstractions;

namespace Sudoku.Tests;

public class ChallengeTests(ITestOutputHelper output)
{
    [Fact]
    public void Evolve_AlwaysOne_Observe()
    {
        var steps = Enumerable.Repeat(1, 100)
            .Aggregate(new List<int>(), (acc, _) =>
            {
                acc.AddRange(new[] { 1, 1, 1, 1, 1 });
                return acc;
            })
            .ToArray();
        var initial = new Challenge.State(0, 0, 0, 0, 0);
        
        var evolution = Challenge.Evolve(initial, steps).ToArray();

        var loopLength = evolution.Length / evolution.Count(s => s == evolution[0]);
        output.WriteLine($"loop size is {loopLength}");

        foreach (var (state, index) in evolution.Select((s,i) => (s,i))) output.WriteLine($"{index+1} {state}");

        evolution.Contains(new Challenge.State(4, 4, 4, 4, 4)).Should().BeTrue();
    }

    [Fact]
    public void BugOnCube_MinimalPathBetweenDiagonalVertices()
    {
        const int scale = 100;
        var minDistanceData = Enumerable.Range(0, 45*scale)
            .AsParallel()
            .WithExecutionMode(ParallelExecutionMode.ForceParallelism)
            .WithDegreeOfParallelism(Environment.ProcessorCount-1)
            .Select(grad => (grad: grad*1.0/scale, rad: grad*1.0/scale * Math.PI / 180))
            .Select(item => (item.grad, 1 / Math.Cos(item.rad) + 1 / Math.Cos(Math.PI / 4 - item.rad)))
            .MinBy(item => item.Item2);

        output.WriteLine($"min distance is {minDistanceData.Item2} for alpha {minDistanceData.grad}");
    }

    [Fact]
    [UseReporter(typeof(VisualStudioReporter), typeof(XUnit2Reporter))]
    public void GenerateStatistics_DifferentTotals_ObserveOptimal()
    {
        var statistics = Enumerable
            .Range(9, 81) //from 9 to 90
            .Select(n => LinksInTeams
                .GenerateStatistics(n)
                .MinBy(s => s.mLinks))
            .ToArray();

        Approvals.VerifyAll(statistics,
            item => $"{item.n}->{item.nLinks} vs {item.m}->{item.mLinks};" +
                    $" {item.n % item.m} by {item.n / item.m + 1}" +
                    $" and {item.m - item.n % item.m} by {item.n / item.m}");

        foreach (var item in statistics)
        {
            output.WriteLine($"{item.n}->{item.nLinks} vs {item.m}->{item.mLinks};" +
                             $" {item.n % item.m} by {item.n / item.m + 1}" +
                             $" and {item.m - item.n % item.m} by {item.n / item.m}");
        }
    }

    [Fact]
    public void GenerateStatistics_DifferentTotals_FindMaxGroupSize()
    {
        var r = Enumerable
            .Range(10, 10_000) //from 10 to 10_010
            .AsParallel()
            .WithExecutionMode(ParallelExecutionMode.ForceParallelism)
            .WithDegreeOfParallelism(Environment.ProcessorCount - 1)
            .Select(n => LinksInTeams
                .GenerateStatistics(n)
                .MinBy(s => s.mLinks))
            .MaxBy(s => s.n/s.m + 1)
            ;

        output.WriteLine($"case of the biggest group is" +
                         $"{r.n} -> {r.nLinks} vs {r.m} -> {r.mLinks}" +
                         $" of {r.n % r.m} by {r.n / r.m + 1}" +
                         $" and {r.m - r.n % r.m} by {r.n / r.m}");
    }
}