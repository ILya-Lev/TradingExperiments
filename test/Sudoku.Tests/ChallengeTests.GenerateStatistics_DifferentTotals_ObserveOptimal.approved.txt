9->36 vs 3->12; 0 by 4 and 3 by 3
10->45 vs 4->14; 2 by 3 and 2 by 2
11->55 vs 4->16; 3 by 3 and 1 by 2
12->66 vs 4->18; 0 by 4 and 4 by 3
13->78 vs 4->21; 1 by 4 and 3 by 3
14->91 vs 5->23; 4 by 3 and 1 by 2
15->105 vs 5->25; 0 by 4 and 5 by 3
16->120 vs 5->28; 1 by 4 and 4 by 3
17->136 vs 5->31; 2 by 4 and 3 by 3
18->153 vs 6->33; 0 by 4 and 6 by 3
19->171 vs 6->36; 1 by 4 and 5 by 3
20->190 vs 6->39; 2 by 4 and 4 by 3
21->210 vs 6->42; 3 by 4 and 3 by 3
22->231 vs 6->45; 4 by 4 and 2 by 3
23->253 vs 6->48; 5 by 4 and 1 by 3
24->276 vs 6->51; 0 by 5 and 6 by 4
25->300 vs 7->54; 4 by 4 and 3 by 3
26->325 vs 7->57; 5 by 4 and 2 by 3
27->351 vs 7->60; 6 by 4 and 1 by 3
28->378 vs 7->63; 0 by 5 and 7 by 4
29->406 vs 7->67; 1 by 5 and 6 by 4
30->435 vs 8->70; 6 by 4 and 2 by 3
31->465 vs 8->73; 7 by 4 and 1 by 3
32->496 vs 8->76; 0 by 5 and 8 by 4
33->528 vs 8->80; 1 by 5 and 7 by 4
34->561 vs 8->84; 2 by 5 and 6 by 4
35->595 vs 9->87; 8 by 4 and 1 by 3
36->630 vs 9->90; 0 by 5 and 9 by 4
37->666 vs 9->94; 1 by 5 and 8 by 4
38->703 vs 9->98; 2 by 5 and 7 by 4
39->741 vs 9->102; 3 by 5 and 6 by 4
40->780 vs 10->105; 0 by 5 and 10 by 4
41->820 vs 10->109; 1 by 5 and 9 by 4
42->861 vs 10->113; 2 by 5 and 8 by 4
43->903 vs 10->117; 3 by 5 and 7 by 4
44->946 vs 10->121; 4 by 5 and 6 by 4
45->990 vs 10->125; 5 by 5 and 5 by 4
46->1035 vs 10->129; 6 by 5 and 4 by 4
47->1081 vs 10->133; 7 by 5 and 3 by 4
48->1128 vs 10->137; 8 by 5 and 2 by 4
49->1176 vs 10->141; 9 by 5 and 1 by 4
50->1225 vs 10->145; 0 by 6 and 10 by 5
51->1275 vs 11->149; 7 by 5 and 4 by 4
52->1326 vs 11->153; 8 by 5 and 3 by 4
53->1378 vs 11->157; 9 by 5 and 2 by 4
54->1431 vs 11->161; 10 by 5 and 1 by 4
55->1485 vs 11->165; 0 by 6 and 11 by 5
56->1540 vs 11->170; 1 by 6 and 10 by 5
57->1596 vs 12->174; 9 by 5 and 3 by 4
58->1653 vs 12->178; 10 by 5 and 2 by 4
59->1711 vs 12->182; 11 by 5 and 1 by 4
60->1770 vs 12->186; 0 by 6 and 12 by 5
61->1830 vs 12->191; 1 by 6 and 11 by 5
62->1891 vs 12->196; 2 by 6 and 10 by 5
63->1953 vs 13->200; 11 by 5 and 2 by 4
64->2016 vs 13->204; 12 by 5 and 1 by 4
65->2080 vs 13->208; 0 by 6 and 13 by 5
66->2145 vs 13->213; 1 by 6 and 12 by 5
67->2211 vs 13->218; 2 by 6 and 11 by 5
68->2278 vs 13->223; 3 by 6 and 10 by 5
69->2346 vs 14->227; 13 by 5 and 1 by 4
70->2415 vs 14->231; 0 by 6 and 14 by 5
71->2485 vs 14->236; 1 by 6 and 13 by 5
72->2556 vs 14->241; 2 by 6 and 12 by 5
73->2628 vs 14->246; 3 by 6 and 11 by 5
74->2701 vs 14->251; 4 by 6 and 10 by 5
75->2775 vs 15->255; 0 by 6 and 15 by 5
76->2850 vs 15->260; 1 by 6 and 14 by 5
77->2926 vs 15->265; 2 by 6 and 13 by 5
78->3003 vs 15->270; 3 by 6 and 12 by 5
79->3081 vs 15->275; 4 by 6 and 11 by 5
80->3160 vs 15->280; 5 by 6 and 10 by 5
81->3240 vs 15->285; 6 by 6 and 9 by 5
82->3321 vs 15->290; 7 by 6 and 8 by 5
83->3403 vs 15->295; 8 by 6 and 7 by 5
84->3486 vs 15->300; 9 by 6 and 6 by 5
85->3570 vs 15->305; 10 by 6 and 5 by 5
86->3655 vs 15->310; 11 by 6 and 4 by 5
87->3741 vs 15->315; 12 by 6 and 3 by 5
88->3828 vs 15->320; 13 by 6 and 2 by 5
89->3916 vs 15->325; 14 by 6 and 1 by 5
